src/lib/replication/repl_dev/raft_repl_dev.cpp

```diff
@@ -356,154 +255,205 @@ void RaftReplDev::on_push_data_received(intrusive< sisl::GenericRpcData >& rpc_d
+    if (!rreq->has_linked_data()) { return rreq; }
+
     // We need to allocate the block, since entry doesn't exist or if it exist, two threads are trying to do the same
     // thing. So take state mutex and allocate the blk
-    std::unique_lock< std::mutex > lg(rreq->state_mtx);
-    if (rreq->state.load() & uint32_cast(repl_req_state_t::BLK_ALLOCATED)) { return rreq; }
+    std::unique_lock< std::mutex > lg(rreq->m_state_mtx);
+    if (rreq->has_state(repl_req_state_t::BLK_ALLOCATED)) { return rreq; }
 
-    auto hints_result = m_listener->get_blk_alloc_hints(user_header, data_size);
+    auto alloc_status = rreq->alloc_local_blks(m_listener, data_size);
 #ifdef _PRERELEASE
     if (is_data_channel) {
         if (iomgr_flip::instance()->test_flip("fake_reject_append_data_channel")) {
             LOGINFO("Data Channel: Reject append_entries flip is triggered for rkey={}", rkey.to_string());
-            hints_result = folly::makeUnexpected(ReplServiceError::NO_SPACE_LEFT);
+            alloc_status = ReplServiceError::NO_SPACE_LEFT;
         }
     } else {
         if (iomgr_flip::instance()->test_flip("fake_reject_append_raft_channel")) {
             LOGINFO("Raft Channel: Reject append_entries flip is triggered for rkey={}", rkey.to_string());
-            hints_result = folly::makeUnexpected(ReplServiceError::NO_SPACE_LEFT);
+            alloc_status = ReplServiceError::NO_SPACE_LEFT;
         }
     }
 #endif
 
-    if (hints_result.hasError()) {
-        RD_LOG(ERROR, "For Repl_key=[{}] alloc hints returned error={}, failing this req", rkey.to_string(),
-               hints_result.error());
+    if (alloc_status != ReplServiceError::OK) {
+        RD_LOGE("For Repl_key=[{}] alloc hints returned error={}, failing this req", rkey.to_string(), alloc_status);
         // Do not call handle_error here, because handle_error is for rreq which needs to be terminated. This one can be
         // retried.
         return nullptr;
     }
 
-    auto const status = data_service().alloc_blks(sisl::round_up(data_size, data_service().get_blk_size()),
-                                                  hints_result.value(), rreq->local_blkid);
-    if (status != BlkAllocStatus::SUCCESS) {
-        RD_LOG(ERROR, "For Repl_key=[{}] allocation failed with status={}, failing this req", status);
-        return nullptr;
+    RD_LOGD("in follower_create_req: rreq={}, addr={}", rreq->to_compact_string(),
+            reinterpret_cast< uintptr_t >(rreq.get()));
+    return rreq;
+}
+
+folly::Future< folly::Unit > RaftReplDev::notify_after_data_written(std::vector< repl_req_ptr_t >* rreqs) {
+    std::vector< folly::Future< folly::Unit > > futs;
+    futs.reserve(rreqs->size());
+    std::vector< repl_req_ptr_t > unreceived_data_reqs;
+
+    // Walk through the list of requests and wait for the data to be received and written
+    for (auto const& rreq : *rreqs) {
+        if ((rreq == nullptr) || (!rreq->has_linked_data())) { continue; }
```

Comment:
When will rreq be nullptr? it is retrieved from `ReplLogStore::end_of_append_batch` via `m_sm.lsn_to_req`, so after a log entry is appended and it can happen that some lsn doesn't map to any rreq, is it a reasonable case we should accept?

Reply:
rreq can be nullptr on cases where this is an internal raft entry. end_of_append_batch contains lsns which are user and raft config as well. Having said that I changed that in end_of_append_batch to not put that requests into vector, so notify_after_data_written will not contain null rreq_ptrs.